glib/src/boxed_any_object.rs

```diff
@@ -0,0 +1,164 @@
+        fn default() -> Self {
+            Self {
+                value: RefCell::new(Box::new(None::<usize>)),
+            }
+        }
+    }
+    impl ObjectImpl for BoxedAnyObject {}
+}
+
+wrapper! {
+    /// This is a type subclassing `glib::object::Object` capable of storing any Rust type.
+    /// It let's you insert a Rust type anywhere a `glib::object::Object` is needed.
+    /// The inserted value can then be borrowed as a Rust type, by using the various
+    /// included methods.
+    ///
+    /// # Examples
+    /// ```
+    /// use glib::BoxedAnyObject;
+    /// struct Author {
+    ///     name: String,
+    ///     subscribers: usize
+    /// }
+    /// // BoxedAnyObject can contain any custom type
+    /// let boxed = BoxedAnyObject::new(Author {
+    ///     name: String::from("GLibAuthor"),
+    ///     subscribers: 1000
+    /// });
+    ///
+    /// // The boxed data can be stored as a `glib::object::Object`
+    /// let list = gio::ListStore::new(BoxedAnyObject::static_type());
+    /// list.append(boxed.clone().upcast());
+    ///
+    /// // And can be retrieved with `borrow`
+    /// let author: Author = boxed.borrow();
+    /// ```
+    pub struct BoxedAnyObject(ObjectSubclass<imp::BoxedAnyObject>);
+}
+
+impl BoxedAnyObject {
+    pub fn new<T: 'static>(value: T) -> Self {
+        let obj: Self = Object::new(&[]).expect("Failed to create BoxedAnyObject");
+        obj.replace(value);
+        obj
+    }
+
+    /// Replaces the wrapped value with a new one, returning the old value, without deinitializing either one.
+    /// The returned value is inside a Box and must be manually downcasted.
+    pub fn replace<T: 'static>(&self, t: T) -> Box<dyn Any> {
+        self.impl_().value.replace(Box::new(t) as Box<dyn Any>)
+    }
+
+    /// Immutably borrows the wrapped value, returning an error if the value is currently mutably
+    /// borrowed or if it's not of type T.
+    ///
+    /// The borrow lasts until the returned `Ref` exits scope. Multiple immutable borrows can be
+    /// taken out at the same time.
+    ///
+    /// This is the non-panicking variant of [`borrow`](#method.borrow).
+    pub fn try_borrow<T: 'static>(&self) -> Result<Ref<'_, T>, BorrowError> {
+        // The required function is only available on nightly:
+        // https://doc.rust-lang.org/std/cell/struct.Ref.html#method.filter_map.
+        // As a workaround, I check if everything is safe, then I unwrap
+
+        let borrowed = self.impl_().value.try_borrow()?;
+        borrowed
+            .as_ref()
+            .downcast_ref::<T>()
+            .ok_or(BorrowError::InvalidType)?;
+        Ok(self.borrow()) // Now this won't panic
+    }
+
+    /// Mutably borrows the wrapped value, returning an error if the value is currently borrowed.
+    /// or if it's not of type T.
+    ///
+    /// The borrow lasts until the returned `RefMut` or all `RefMut`s derived
+    /// from it exit scope. The value cannot be borrowed while this borrow is
```

Comment:
```suggestion
    /// from it exit scope. The value cannot be borrowed another time while this borrow is
```

Reply:
I'm not gonna change this phrase, since it's copied from `std::cell::RefMut` xD

Reply:
Ok :)